// @fluentui/react-avatar/lib-commonjs/components/AvatarGroupItem/useAvatarGroupItem.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useAvatarGroupItemBase_unstable: function() {
        return useAvatarGroupItemBase_unstable;
    },
    useAvatarGroupItem_unstable: function() {
        return useAvatarGroupItem_unstable;
    }
});
const _Avatar = require("../Avatar/Avatar");
const _AvatarGroupContext = require("../../contexts/AvatarGroupContext");
const _useAvatarGroup = require("../AvatarGroup/useAvatarGroup");
const _reactutilities = require("@fluentui/react-utilities");
const _reactcontextselector = require("@fluentui/react-context-selector");
const useAvatarGroupItem_unstable = (props, ref)=>{
    const state = useAvatarGroupItemBase_unstable(props, ref);
    const groupSize = (0, _AvatarGroupContext.useAvatarGroupContext_unstable)((ctx)=>ctx.size);
    const size = groupSize !== null && groupSize !== void 0 ? groupSize : _useAvatarGroup.defaultAvatarGroupSize;
    return {
        size,
        ...state,
        components: {
            // eslint-disable-next-line @typescript-eslint/no-deprecated
            ...state.components,
            avatar: _Avatar.Avatar
        },
        avatar: _reactutilities.slot.always(props.avatar, {
            defaultProps: {
                size,
                color: 'colorful',
                ...state.avatar
            },
            elementType: _Avatar.Avatar
        })
    };
};
const useAvatarGroupItemBase_unstable = (props, ref)=>{
    const groupIsOverflow = (0, _AvatarGroupContext.useAvatarGroupContext_unstable)((ctx)=>ctx.isOverflow);
    const layout = (0, _AvatarGroupContext.useAvatarGroupContext_unstable)((ctx)=>ctx.layout);
    // Since the primary slot is not an intrinsic element, getPartitionedNativeProps cannot be used here.
    const { style, className, overflowLabel, ...avatarSlotProps } = props;
    const hasAvatarGroupContext = (0, _reactcontextselector.useHasParentContext)(_AvatarGroupContext.AvatarGroupContext);
    if (process.env.NODE_ENV !== 'production' && !hasAvatarGroupContext) {
        // eslint-disable-next-line no-console
        console.warn('AvatarGroupItem must only be used inside an AvatarGroup component.');
    }
    return {
        isOverflowItem: groupIsOverflow,
        layout,
        components: {
            root: groupIsOverflow ? 'li' : 'div',
            avatar: 'span',
            overflowLabel: 'span'
        },
        root: _reactutilities.slot.always(props.root, {
            defaultProps: {
                style,
                className
            },
            elementType: groupIsOverflow ? 'li' : 'div'
        }),
        avatar: _reactutilities.slot.always(props.avatar, {
            defaultProps: {
                ref,
                ...avatarSlotProps
            },
            elementType: 'span'
        }),
        overflowLabel: _reactutilities.slot.always(overflowLabel, {
            defaultProps: {
                // Avatar already has its aria-label set to the name, this will prevent the name to be read twice.
                'aria-hidden': true,
                children: props.name
            },
            elementType: 'span'
        })
    };
};
