// @fluentui/react-badge/lib-commonjs/components/Badge/useBadge.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useBadgeBase_unstable: function() {
        return useBadgeBase_unstable;
    },
    useBadge_unstable: function() {
        return useBadge_unstable;
    }
});
const _reactutilities = require("@fluentui/react-utilities");
const useBadge_unstable = (props, ref)=>{
    const { shape = 'circular', size = 'medium', appearance = 'filled', color = 'brand', ...badgeProps } = props;
    const state = useBadgeBase_unstable(badgeProps, ref);
    return {
        ...state,
        shape,
        size,
        appearance,
        color
    };
};
const useBadgeBase_unstable = (props, ref)=>{
    const { iconPosition = 'before' } = props;
    return {
        iconPosition,
        components: {
            root: 'div',
            icon: 'span'
        },
        root: _reactutilities.slot.always((0, _reactutilities.getIntrinsicElementProps)('div', {
            ref,
            ...props
        }), {
            elementType: 'div'
        }),
        icon: _reactutilities.slot.optional(props.icon, {
            elementType: 'span'
        })
    };
};
