// @fluentui/react-motion-components-preview/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Blur: function() {
        return _Blur.Blur;
    },
    Collapse: function() {
        return _Collapse.Collapse;
    },
    CollapseDelayed: function() {
        return _Collapse.CollapseDelayed;
    },
    CollapseRelaxed: function() {
        return _Collapse.CollapseRelaxed;
    },
    CollapseSnappy: function() {
        return _Collapse.CollapseSnappy;
    },
    Fade: function() {
        return _Fade.Fade;
    },
    FadeRelaxed: function() {
        return _Fade.FadeRelaxed;
    },
    FadeSnappy: function() {
        return _Fade.FadeSnappy;
    },
    Rotate: function() {
        return _Rotate.Rotate;
    },
    Scale: function() {
        return _Scale.Scale;
    },
    ScaleRelaxed: function() {
        return _Scale.ScaleRelaxed;
    },
    ScaleSnappy: function() {
        return _Scale.ScaleSnappy;
    },
    Slide: function() {
        return _Slide.Slide;
    },
    SlideRelaxed: function() {
        return _Slide.SlideRelaxed;
    },
    SlideSnappy: function() {
        return _Slide.SlideSnappy;
    },
    Stagger: function() {
        return _Stagger.Stagger;
    },
    blurAtom: function() {
        return _bluratom.blurAtom;
    },
    fadeAtom: function() {
        return _fadeatom.fadeAtom;
    },
    rotateAtom: function() {
        return _rotateatom.rotateAtom;
    },
    scaleAtom: function() {
        return _scaleatom.scaleAtom;
    },
    slideAtom: function() {
        return _slideatom.slideAtom;
    }
});
const _Collapse = require("./components/Collapse");
const _Fade = require("./components/Fade");
const _Scale = require("./components/Scale");
const _Slide = require("./components/Slide");
const _Blur = require("./components/Blur");
const _Rotate = require("./components/Rotate");
const _Stagger = require("./choreography/Stagger");
const _bluratom = require("./atoms/blur-atom");
const _fadeatom = require("./atoms/fade-atom");
const _rotateatom = require("./atoms/rotate-atom");
const _scaleatom = require("./atoms/scale-atom");
const _slideatom = require("./atoms/slide-atom");
 // TODO: consider whether to export some or all collapse atoms
