// @apolo/shared-api/src/client.js
/**
 * Factory que crea un cliente API de Apolo con el contexto Xrm inyectado.
 * En producción se pasa el Xrm real; en tests/dev se puede pasar un mock.
 *
 * @param {Object} xrmContext - Objeto Xrm de Power Platform
 * @returns {{ odata, customApi, decrypt, retrieveById, retrieveByIdExpand }}
 */
export const createApoloClient = (xrmContext) => {
  const getClientUrl = () => xrmContext.Utility.getGlobalContext().getClientUrl()

  /**
   * Ejecuta una query OData v9.1
   */
  const odata = async (entityName, filter, method = 'GET') => {
    try {
      const url = getClientUrl()
      const response = await fetch(`${url}/api/data/v9.1/${entityName}?${filter}`, {
        method,
        headers: {
          'OData-MaxVersion': '4.0',
          'OData-Version': '4.0',
          'Content-Type': 'application/json; charset=utf-8',
          Accept: 'application/json',
          Prefer: 'odata.include-annotations=*',
        },
      })
      const data = await response.json()
      return data.value?.length > 0 ? data.value : null
    } catch {
      return null
    }
  }

  /**
   * Llama a una Custom API v9.2
   */
  const customApi = async (method, parameters, endpointName) => {
    try {
      const url = getClientUrl()
      const response = await fetch(`${url}/api/data/v9.2/${endpointName}`, {
        method,
        headers: {
          'OData-MaxVersion': '4.0',
          'OData-Version': '4.0',
          'Content-Type': 'application/json; charset=utf-8',
          Accept: 'application/json',
        },
        body: JSON.stringify(parameters),
      })
      return await response.json()
    } catch {
      return { Error: `Error al llamar a custom api ${endpointName}`, Response: null }
    }
  }

  /**
   * Desencripta un dato usando el servicio AES-GCM
   */
  const decrypt = async (dato) => {
    try {
      const { customApis } = await import('./config/endpoints.js')
      const { configuracionesEncryptDecrypt: enc } = await import('./config/encrypt.js')
      const url = getClientUrl()
      const parameters = {
        EncryptDescryptApi_In_rfsc_text_to_Desencrypt: dato,
        EncryptDescryptApi_In_key: enc.AESGCM,
      }
      const response = await fetch(`${url}/api/data/v9.2/${customApis.DesencriptarDatos}`, {
        method: 'POST',
        headers: {
          'OData-MaxVersion': '4.0',
          'OData-Version': '4.0',
          'Content-Type': 'application/json; charset=utf-8',
          Accept: 'application/json',
        },
        body: JSON.stringify(parameters),
      })
      const data = await response.json()
      return data.EncryptDescryptApi_Out_rfsc_text_to_Desencrypt
    } catch {
      return ''
    }
  }

  /**
   * Obtiene un registro por ID usando Xrm.WebApi
   */
  const retrieveById = async (entityName, id) => {
    try {
      return await xrmContext.WebApi.online.retrieveRecord(entityName, id)
    } catch {
      return null
    }
  }

  /**
   * Obtiene un registro por ID con $expand
   */
  const retrieveByIdExpand = async (entityName, id, expandList) => {
    try {
      let expand = '?$expand='
      for (const item of expandList) {
        expand +=
          expand === '?$expand='
            ? `${item.entityExpand}($select=${item.selectExpand})`
            : `,${item.entityExpand}($select=${item.selectExpand})`
      }
      return await xrmContext.WebApi.online.retrieveRecord(entityName, id, expand)
    } catch {
      return null
    }
  }

  return { odata, customApi, decrypt, retrieveById, retrieveByIdExpand }
}
