// @griffel/core/lib/shorthands/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get border () {
        return _border.border;
    },
    get borderBottom () {
        return _borderBottom.borderBottom;
    },
    get borderColor () {
        return _borderColor.borderColor;
    },
    get borderLeft () {
        return _borderLeft.borderLeft;
    },
    get borderRadius () {
        return _borderRadius.borderRadius;
    },
    get borderRight () {
        return _borderRight.borderRight;
    },
    get borderStyle () {
        return _borderStyle.borderStyle;
    },
    get borderTop () {
        return _borderTop.borderTop;
    },
    get borderWidth () {
        return _borderWidth.borderWidth;
    },
    get flex () {
        return _flex.flex;
    },
    get gap () {
        return _gap.gap;
    },
    get gridArea () {
        return _gridArea.gridArea;
    },
    get inset () {
        return _inset.inset;
    },
    get margin () {
        return _margin.margin;
    },
    get marginBlock () {
        return _marginBlock.marginBlock;
    },
    get marginInline () {
        return _marginInline.marginInline;
    },
    get outline () {
        return _outline.outline;
    },
    get overflow () {
        return _overflow.overflow;
    },
    get padding () {
        return _padding.padding;
    },
    get paddingBlock () {
        return _paddingBlock.paddingBlock;
    },
    get paddingInline () {
        return _paddingInline.paddingInline;
    },
    get textDecoration () {
        return _textDecoration.textDecoration;
    },
    get transition () {
        return _transition.transition;
    }
});
const _border = require("./border.cjs");
const _borderLeft = require("./borderLeft.cjs");
const _borderBottom = require("./borderBottom.cjs");
const _borderRight = require("./borderRight.cjs");
const _borderTop = require("./borderTop.cjs");
const _borderColor = require("./borderColor.cjs");
const _borderStyle = require("./borderStyle.cjs");
const _borderRadius = require("./borderRadius.cjs");
const _borderWidth = require("./borderWidth.cjs");
const _flex = require("./flex.cjs");
const _gap = require("./gap.cjs");
const _gridArea = require("./gridArea.cjs");
const _margin = require("./margin.cjs");
const _marginBlock = require("./marginBlock.cjs");
const _marginInline = require("./marginInline.cjs");
const _padding = require("./padding.cjs");
const _paddingBlock = require("./paddingBlock.cjs");
const _paddingInline = require("./paddingInline.cjs");
const _overflow = require("./overflow.cjs");
const _inset = require("./inset.cjs");
const _outline = require("./outline.cjs");
const _transition = require("./transition.cjs");
const _textDecoration = require("./textDecoration.cjs");
