// @fluentui/react-tags/lib-commonjs/components/InteractionTagPrimary/useInteractionTagPrimary.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useInteractionTagPrimaryBase_unstable: function() {
        return useInteractionTagPrimaryBase_unstable;
    },
    useInteractionTagPrimary_unstable: function() {
        return useInteractionTagPrimary_unstable;
    }
});
const _reactutilities = require("@fluentui/react-utilities");
const _interactionTagContext = require("../../contexts/interactionTagContext");
const avatarSizeMap = {
    medium: 28,
    small: 20,
    'extra-small': 16
};
const avatarShapeMap = {
    rounded: 'square',
    circular: 'circular'
};
const useInteractionTagPrimaryBase_unstable = (props, ref)=>{
    const { disabled, handleTagSelect, interactionTagPrimaryId, selected: contextSelected, selectedValues, value } = (0, _interactionTagContext.useInteractionTagContext_unstable)();
    const { hasSecondaryAction = false } = props;
    const onClick = (0, _reactutilities.useEventCallback)((0, _reactutilities.mergeCallbacks)(props === null || props === void 0 ? void 0 : props.onClick, (event)=>handleTagSelect === null || handleTagSelect === void 0 ? void 0 : handleTagSelect(event, {
            type: 'click',
            event,
            value,
            selectedValues
        })));
    return {
        disabled,
        hasSecondaryAction,
        selected: contextSelected,
        components: {
            root: 'button',
            media: 'span',
            icon: 'span',
            primaryText: 'span',
            secondaryText: 'span'
        },
        root: _reactutilities.slot.always((0, _reactutilities.getIntrinsicElementProps)('button', {
            ref,
            disabled,
            id: interactionTagPrimaryId,
            ...handleTagSelect !== undefined && {
                'aria-pressed': contextSelected
            },
            onClick,
            ...props
        }), {
            elementType: 'button'
        }),
        media: _reactutilities.slot.optional(props.media, {
            elementType: 'span'
        }),
        icon: _reactutilities.slot.optional(props.icon, {
            elementType: 'span'
        }),
        primaryText: _reactutilities.slot.optional(props.primaryText, {
            renderByDefault: true,
            defaultProps: {
                children: props.children
            },
            elementType: 'span'
        }),
        secondaryText: _reactutilities.slot.optional(props.secondaryText, {
            elementType: 'span'
        })
    };
};
const useInteractionTagPrimary_unstable = (props, ref)=>{
    const { appearance, shape, size } = (0, _interactionTagContext.useInteractionTagContext_unstable)();
    return {
        ...useInteractionTagPrimaryBase_unstable(props, ref),
        appearance,
        avatarShape: avatarShapeMap[shape],
        avatarSize: avatarSizeMap[size],
        shape,
        size
    };
};
