// @griffel/core/src/common/snapshotSerializers.js
import * as prettier from 'prettier';
import { DATA_BUCKET_ATTR } from '../constants.js';
import { normalizeCSSBucketEntry } from '../runtime/utils/normalizeCSSBucketEntry.js';
// eslint-disable-next-line eqeqeq
const isObject = (value) => value != null && !Array.isArray(value) && typeof value === 'object';
export const griffelRendererSerializer = {
    test(value) {
        return isObject(value) && isObject(value.stylesheets);
    },
    print(value) {
        /**
         * test function makes sure that value is the guarded type
         */
        const _value = value;
        const stylesheetKeys = Object.keys(_value.stylesheets);
        const rules = stylesheetKeys.reduce((acc, styleEl) => {
            var _a;
            const stylesheet = _value.stylesheets[styleEl];
            if (stylesheet) {
                const cssRules = (_a = stylesheet.cssRules()) !== null && _a !== void 0 ? _a : [];
                const attributes = Object.entries(stylesheet.elementAttributes).filter(([key]) => key !== DATA_BUCKET_ATTR);
                if (cssRules.length === 0) {
                    return acc;
                }
                return [
                    ...acc,
                    `/** bucket "${styleEl.slice(0, 1)}"${attributes.length > 0 ? ' ' + JSON.stringify(Object.fromEntries(attributes)) : ''} **/`,
                    ...cssRules,
                ];
            }
            return acc;
        }, []);
        return prettier.format(rules.join('\n'), { parser: 'css' }).trim();
    },
};
const isStyleRulesTuple = (value) => Array.isArray(value) && value.length === 2;
export const griffelRulesSerializer = {
    test(value) {
        return isStyleRulesTuple(value);
    },
    print(value) {
        /**
         * test function makes sure that value is the guarded type
         */
        const _value = value;
        const cssRulesByBucket = _value[1];
        const keys = Object.keys(cssRulesByBucket);
        return keys.reduce((acc, styleBucketName) => {
            const rules = cssRulesByBucket[styleBucketName].map(entry => normalizeCSSBucketEntry(entry)[0]);
            return prettier.format(acc + rules.join(''), { parser: 'css' }).trim();
        }, '');
    },
};
export const griffelResetRulesSerializer = {
    test(value) {
        return Array.isArray(value);
    },
    print(value) {
        const format = (cssRules) => prettier.format(cssRules.join(''), { parser: 'css' }).trim();
        /**
         * test function makes sure that value is the guarded type
         */
        const _value = value;
        const cssRulesByBucket = Array.isArray(_value[2]) ? { r: _value[2] } : _value[2];
        return Object.entries(cssRulesByBucket)
            .filter(([_key, value]) => value.length > 0)
            .flatMap(([key, value]) => [`/** bucket "${key}" */`, format(value)])
            .join('\n');
    },
};
