// @apolo/shared-api/src/config/operation-codes.js
export const codigosMatrizOperaciones = {
  bloqueo: '2200',
  desbloqueo: '2201',
  reemision: '1085',
  embozado: '2061',
  firma_contrato: 'rw_firma_contrato',
  prepago: '2062',
  certificados_deudaConsolidada: '1069',
}

export const configuracionesOperaciones = {
  reemision: {
    inicio: true,
    listar: { operacion: '1084', permiso: false, mensaje: 'No cuentas con permisos para listar reemisiones' },
    ejecutar: { operacion: '1085', permiso: false, mensaje: 'No cuentas con permisos para ejecutar reemisiones' },
  },
  bloqueo_tarjeta: {
    inicio: true,
    ejecutar: [
      { codigo: 'TA02S', operacion: '1081', descripcion: 'Bloqueo_Tarjeta_Emisior', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por emisor' },
      { codigo: 'TA02O', operacion: '1082', descripcion: 'Bloqueo_Tarjeta_Perdida', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por perdida' },
      { codigo: 'TA02A', operacion: '1083', descripcion: 'bloqueo_Tarjeta_Fallecido_Adicional', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por fallecido adicional' },
      { codigo: 'TA02F', operacion: '1086', descripcion: 'bloqueo_Tarjeta_Fraude', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por fraude' },
      { codigo: 'TA02K', operacion: '1088', descripcion: 'Bloqueo_Tarjeta_Renuncia_Tarjeta', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por renuncia tarjeta' },
      { codigo: 'TA02Q', operacion: '1079', descripcion: 'Bloqueo_Tarjeta_Administrativo_Tarjeta', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por administrativo tarjeta' },
      { codigo: 'TA022', operacion: '2111', descripcion: 'Bloqueo_Tarjeta_Inactividad', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por inactividad' },
      { codigo: 'TA02L', operacion: '2112', descripcion: 'Bloqueo_Tarjeta_ON_OFF', permiso: false, mensaje: 'No cuentas con permisos para bloquear tarjeta por on off' },
    ],
  },
  bloqueo_cuenta: {
    inicio: true,
    ejecutar: [
      { codigo: 'CA02Q', operacion: '1079', descripcion: 'Bloqueo_Cuenta_Adm_Tarjeta', permiso: false, mensaje: 'No cuentas con permisos para bloquear la cuenta por administración de tarjeta' },
      { codigo: 'CA02A', operacion: '1074', descripcion: 'Bloqueo_Cuenta_Fallecido', permiso: false, mensaje: 'No cuentas con permisos para bloquear la cuenta por fallecido' },
      { codigo: 'CA02F', operacion: '1075', descripcion: 'Bloqueo_Cuenta_Fraude', permiso: false, mensaje: 'No cuentas con permisos para bloquear la cuenta por fraude' },
      { codigo: 'CA02V', operacion: '1076', descripcion: 'Bloqueo_Cuenta_Renuncia_Cta', permiso: false, mensaje: 'No cuentas con permisos para bloquear la cuenta por renuncia cuenta' },
    ],
  },
  desbloqueo_cuenta: {
    inicio: true,
    ejecutar: [
      { operacion: '1080', codigo: 'CA02Q', descripcion: 'Desbloqueo_Cuenta_Adm_Tarjeta', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la cuenta por administración de tarjeta' },
      { operacion: '2065', codigo: 'CA02M', descripcion: 'Desbloqueo_Cuenta_RDR', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la cuenta por RDR' },
      { operacion: '2067', codigo: 'CA02B', descripcion: 'Desbloqueo_Cuenta_Gerencia', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la cuenta por Gerencia' },
    ],
  },
  desbloqueo_tarjeta: {
    inicio: true,
    ejecutar: [
      { operacion: '2063', codigo: 'TA02Q', descripcion: 'Desbloqueo_Tarjeta_Adm_Tarjeta', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la tarjeta por administración de tarjeta' },
      { operacion: '2064', codigo: 'TA02L', descripcion: 'Desbloqueo_Tarjeta_ONOFF', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la tarjeta por captura ON/OFF' },
      { operacion: '1089', codigo: 'TA022', descripcion: 'Desbloqueo_Tarjeta_Inactividad', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la tarjeta por Inactividad' },
      { operacion: '2066', codigo: 'TA02M', descripcion: 'Desbloqueo_Tarjeta_RDR', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la tarjeta por RDR' },
      { operacion: '2067', codigo: 'TA02B', descripcion: 'Desbloqueo_Tarjeta_Gerencia', permiso: false, mensaje: 'No cuentas con permisos para desbloquear la tarjeta por Gerencia' },
    ],
  },
  firma_contrato: {
    inicio: true,
    listar: { operacion: 'rw_firma_contrato', permiso: false, mensaje: 'No cuentas con permisos para firmar contrato' },
  },
  prepago: {
    inicio: true,
    listar: { operacion: '2062', permiso: false, mensaje: 'No cuentas con permisos para ingresar a prepagos.' },
  },
  embozado: {
    inicio: true,
    listar: { operacion: '2061', permiso: false, mensaje: 'No cuentas con permisos para ingresar a embozado.' },
  },
  certificados_deudaConsolidada: {
    inicio: true,
    listar: { operacion: '1069', permiso: false, mensaje: 'No cuentas con permisos para generar certificados de deuda consolidada' },
  },
}

export const getConfiguracionesOperaciones = () => configuracionesOperaciones

export const tipos_prepago = { total: 2, parcial: 1 }

export const tipos_documento_prepago = {
  certificado_parcial: 'cp',
  simulacion_parcial: 'sp',
  certificado_total: 'ct',
  simulacion_total: 'st',
}

export const tipos_home = { activos: 1, embozadosVenta: 2 }

export const EstadosCurseventa = {
  No_apto_para_cursar: 803750000,
  Listo_para_cursar: 1,
  Curse_iniciado: 2,
  Curse_en_legado: 3,
  A_recursar: 4,
  Cursado: 5,
  Error_al_cursar: 6,
  Enviar_a_PCE: 7,
  Cursado_por_Excepcion: 8,
}

export const rolesSeguridad = { Exportador: 'Rol de Exportador de Datos' }

export const Vistas = { VistaMasInfo: 'MasInfo', VistaHome: 'Home', VistaMasInfoNoVigente: 'MasInfoNoVigente' }

export const Secciones = {
  SeccionTarjeta: 'Tarjeta',
  SeccionCuenta: 'Cuenta',
  SeccionClienteSinCuenta: 'ClienteSinCuenta',
  SeccionNoVigente: 'NoVigente',
}
