// @fluentui/react-tags/lib-commonjs/components/InteractionTag/useInteractionTag.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useInteractionTagBase_unstable: function() {
        return useInteractionTagBase_unstable;
    },
    useInteractionTag_unstable: function() {
        return useInteractionTag_unstable;
    }
});
const _reactutilities = require("@fluentui/react-utilities");
const _tagGroupContext = require("../../contexts/tagGroupContext");
const useInteractionTagBase_unstable = (props, ref)=>{
    const { handleTagDismiss, handleTagSelect, disabled: contextDisabled, selectedValues = [] } = (0, _tagGroupContext.useTagGroupContext_unstable)();
    const id = (0, _reactutilities.useId)('fui-InteractionTag-', props.id);
    const interactionTagPrimaryId = (0, _reactutilities.useId)('fui-InteractionTagPrimary-');
    const { disabled = false, selected = false, value = id } = props;
    return {
        disabled: contextDisabled ? true : disabled,
        handleTagDismiss,
        handleTagSelect,
        interactionTagPrimaryId,
        selected: selectedValues.includes(value) || selected,
        selectedValues,
        value,
        components: {
            root: 'div'
        },
        root: _reactutilities.slot.always((0, _reactutilities.getIntrinsicElementProps)('div', {
            ref,
            ...props,
            disabled: contextDisabled ? true : disabled,
            id
        }), {
            elementType: 'div'
        })
    };
};
const useInteractionTag_unstable = (props, ref)=>{
    const { size: contextSize, appearance: contextAppearance } = (0, _tagGroupContext.useTagGroupContext_unstable)();
    const { appearance = contextAppearance !== null && contextAppearance !== void 0 ? contextAppearance : 'filled', shape = 'rounded', size = contextSize } = props;
    return {
        ...useInteractionTagBase_unstable(props, ref),
        appearance,
        shape,
        size
    };
};
