// @griffel/core/lib/runtime/resolveStyleRules.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "resolveStyleRules", {
    enumerable: true,
    get: function() {
        return resolveStyleRules;
    }
});
const _hash = /*#__PURE__*/ _interop_require_default(require("@emotion/hash"));
const _core = require("rtl-css-js/core");
const _constants = require("../constants.cjs");
const _compileAtomicCSSRule = require("./compileAtomicCSSRule.cjs");
const _compileKeyframeCSS = require("./compileKeyframeCSS.cjs");
const _getStyleBucketName = require("./getStyleBucketName.cjs");
const _shorthands = require("./shorthands.cjs");
const _generateCombinedMediaQuery = require("./utils/generateCombinedMediaQuery.cjs");
const _hashClassName = require("./utils/hashClassName.cjs");
const _hashPropertyKey = require("./utils/hashPropertyKey.cjs");
const _isContainerQuerySelector = require("./utils/isContainerQuerySelector.cjs");
const _isLayerSelector = require("./utils/isLayerSelector.cjs");
const _isMediaQuerySelector = require("./utils/isMediaQuerySelector.cjs");
const _isNestedSelector = require("./utils/isNestedSelector.cjs");
const _isObject = require("./utils/isObject.cjs");
const _isResetValue = require("./utils/isResetValue.cjs");
const _isScopeSelector = require("./utils/isScopeSelector.cjs");
const _isSupportQuerySelector = require("./utils/isSupportQuerySelector.cjs");
const _normalizeNestedProperty = require("./utils/normalizeNestedProperty.cjs");
const _trimSelector = require("./utils/trimSelector.cjs");
const _logError = require("./warnings/logError.cjs");
const _warnAboutUnresolvedRule = require("./warnings/warnAboutUnresolvedRule.cjs");
const _warnAboutUnsupportedProperties = require("./warnings/warnAboutUnsupportedProperties.cjs");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function getShorthandDefinition(property) {
    return _shorthands.shorthands[property];
}
function computePropertyPriority(shorthand) {
    var _a;
    return (_a = shorthand === null || shorthand === void 0 ? void 0 : shorthand[0]) !== null && _a !== void 0 ? _a : 0;
}
function pushToClassesMap(classesMap, propertyKey, ltrClassname, rtlClassname) {
    classesMap[propertyKey] = rtlClassname ? [
        ltrClassname,
        rtlClassname
    ] : ltrClassname;
}
function createBucketEntry(cssRule, metadata) {
    if (metadata.length > 0) {
        return [
            cssRule,
            Object.fromEntries(metadata)
        ];
    }
    return cssRule;
}
function pushToCSSRules(cssRulesByBucket, styleBucketName, ltrCSS, rtlCSS, media, priority) {
    var _a;
    const metadata = [];
    if (priority !== 0) {
        metadata.push([
            'p',
            priority
        ]);
    }
    if (styleBucketName === 'm' && media) {
        metadata.push([
            'm',
            media
        ]);
    }
    (_a = cssRulesByBucket[styleBucketName]) !== null && _a !== void 0 ? _a : cssRulesByBucket[styleBucketName] = [];
    if (ltrCSS) {
        cssRulesByBucket[styleBucketName].push(createBucketEntry(ltrCSS, metadata));
    }
    if (rtlCSS) {
        cssRulesByBucket[styleBucketName].push(createBucketEntry(rtlCSS, metadata));
    }
}
function resolveStyleRules(styles, classNameHashSalt = '', selectors = [], atRules = {
    container: '',
    layer: '',
    media: '',
    supports: '',
    scope: ''
}, cssClassesMap = {}, cssRulesByBucket = {}, rtlValue) {
    // eslint-disable-next-line guard-for-in
    for(const property in styles){
        // eslint-disable-next-line no-prototype-builtins
        if (_constants.UNSUPPORTED_CSS_PROPERTIES.hasOwnProperty(property)) {
            (0, _warnAboutUnsupportedProperties.warnAboutUnsupportedProperties)(property, styles[property]);
            continue;
        }
        const value = styles[property];
        // eslint-disable-next-line eqeqeq
        if (value == null) {
            continue;
        }
        if ((0, _isResetValue.isResetValue)(value)) {
            const selector = (0, _trimSelector.trimSelector)(selectors.join(''));
            // uniq key based on a hash of property & selector, used for merging later
            const key = (0, _hashPropertyKey.hashPropertyKey)(selector, property, atRules);
            pushToClassesMap(cssClassesMap, key, 0, undefined);
            continue;
        }
        if (typeof value === 'string' || typeof value === 'number') {
            const selector = (0, _trimSelector.trimSelector)(selectors.join(''));
            const shorthand = getShorthandDefinition(property);
            if (shorthand) {
                const shorthandProperties = shorthand[1];
                const shorthandResetStyles = Object.fromEntries(shorthandProperties.map((property)=>[
                        property,
                        _constants.RESET
                    ]));
                resolveStyleRules(shorthandResetStyles, classNameHashSalt, selectors, atRules, cssClassesMap, cssRulesByBucket);
            }
            // uniq key based on a hash of property & selector, used for merging later
            const key = (0, _hashPropertyKey.hashPropertyKey)(selector, property, atRules);
            const className = (0, _hashClassName.hashClassName)({
                value: value.toString(),
                salt: classNameHashSalt,
                selector,
                property
            }, atRules);
            const rtlDefinition = rtlValue && {
                key: property,
                value: rtlValue
            } || (0, _core.convertProperty)(property, value);
            const flippedInRtl = rtlDefinition.key !== property || rtlDefinition.value !== value;
            const rtlClassName = flippedInRtl ? (0, _hashClassName.hashClassName)({
                value: rtlDefinition.value.toString(),
                property: rtlDefinition.key,
                salt: classNameHashSalt,
                selector
            }, atRules) : undefined;
            const rtlCompileOptions = flippedInRtl ? {
                rtlClassName,
                rtlProperty: rtlDefinition.key,
                rtlValue: rtlDefinition.value
            } : undefined;
            const styleBucketName = (0, _getStyleBucketName.getStyleBucketName)(selectors, atRules);
            const [ltrCSS, rtlCSS] = (0, _compileAtomicCSSRule.compileAtomicCSSRule)({
                className,
                selectors,
                property,
                value,
                ...rtlCompileOptions
            }, atRules);
            pushToClassesMap(cssClassesMap, key, className, rtlClassName);
            pushToCSSRules(cssRulesByBucket, styleBucketName, ltrCSS, rtlCSS, atRules.media, computePropertyPriority(shorthand));
        } else if (property === 'animationName') {
            const animationNameValue = Array.isArray(value) ? value : [
                value
            ];
            const animationNames = [];
            const rtlAnimationNames = [];
            for (const keyframeObject of animationNameValue){
                const keyframeCSS = (0, _compileKeyframeCSS.compileKeyframeRule)(keyframeObject);
                const rtlKeyframeCSS = (0, _compileKeyframeCSS.compileKeyframeRule)((0, _core.convert)(keyframeObject));
                const animationName = _constants.HASH_PREFIX + (0, _hash.default)(keyframeCSS);
                let rtlAnimationName;
                const keyframeRules = (0, _compileKeyframeCSS.compileKeyframesCSS)(animationName, keyframeCSS);
                let rtlKeyframeRules = [];
                if (keyframeCSS === rtlKeyframeCSS) {
                    // If CSS for LTR & RTL are same we will re-use animationName from LTR to avoid duplication of rules in output
                    rtlAnimationName = animationName;
                } else {
                    rtlAnimationName = _constants.HASH_PREFIX + (0, _hash.default)(rtlKeyframeCSS);
                    rtlKeyframeRules = (0, _compileKeyframeCSS.compileKeyframesCSS)(rtlAnimationName, rtlKeyframeCSS);
                }
                for(let i = 0; i < keyframeRules.length; i++){
                    pushToCSSRules(cssRulesByBucket, // keyframes styles should be inserted into own bucket
                    'k', keyframeRules[i], rtlKeyframeRules[i], atRules.media, // keyframes always have default priority
                    0);
                }
                animationNames.push(animationName);
                rtlAnimationNames.push(rtlAnimationName);
            }
            resolveStyleRules({
                animationName: animationNames.join(', ')
            }, classNameHashSalt, selectors, atRules, cssClassesMap, cssRulesByBucket, rtlAnimationNames.join(', '));
        } else if (Array.isArray(value)) {
            // not animationName property but array in the value => fallback values
            if (value.length === 0) {
                if (process.env.NODE_ENV !== 'production') {
                    console.warn(`makeStyles(): An empty array was passed as input to "${property}", the property will be omitted in the styles.`);
                }
                continue;
            }
            const selector = (0, _trimSelector.trimSelector)(selectors.join(''));
            const shorthand = getShorthandDefinition(property);
            if (shorthand) {
                const shorthandProperties = shorthand[1];
                const shorthandResetStyles = Object.fromEntries(shorthandProperties.map((property)=>[
                        property,
                        _constants.RESET
                    ]));
                resolveStyleRules(shorthandResetStyles, classNameHashSalt, selectors, atRules, cssClassesMap, cssRulesByBucket);
            }
            const key = (0, _hashPropertyKey.hashPropertyKey)(selector, property, atRules);
            const className = (0, _hashClassName.hashClassName)({
                value: value.map((v)=>(v !== null && v !== void 0 ? v : '').toString()).join(';'),
                salt: classNameHashSalt,
                selector,
                property
            }, atRules);
            const rtlDefinitions = value.map((v)=>(0, _core.convertProperty)(property, v));
            const rtlPropertyConsistent = !rtlDefinitions.some((v)=>v.key !== rtlDefinitions[0].key);
            if (!rtlPropertyConsistent) {
                if (process.env.NODE_ENV !== 'production') {
                    console.error('makeStyles(): mixing CSS fallback values which result in multiple CSS properties in RTL is not supported.');
                }
                continue;
            }
            const flippedInRtl = rtlDefinitions[0].key !== property || rtlDefinitions.some((v, i)=>v.value !== value[i]);
            const rtlClassName = flippedInRtl ? (0, _hashClassName.hashClassName)({
                value: rtlDefinitions.map((v)=>{
                    var _a;
                    return ((_a = v === null || v === void 0 ? void 0 : v.value) !== null && _a !== void 0 ? _a : '').toString();
                }).join(';'),
                salt: classNameHashSalt,
                property: rtlDefinitions[0].key,
                selector
            }, atRules) : undefined;
            const rtlCompileOptions = flippedInRtl ? {
                rtlClassName,
                rtlProperty: rtlDefinitions[0].key,
                rtlValue: rtlDefinitions.map((d)=>d.value)
            } : undefined;
            const styleBucketName = (0, _getStyleBucketName.getStyleBucketName)(selectors, atRules);
            const [ltrCSS, rtlCSS] = (0, _compileAtomicCSSRule.compileAtomicCSSRule)({
                className,
                selectors,
                property,
                value: value,
                ...rtlCompileOptions
            }, atRules);
            pushToClassesMap(cssClassesMap, key, className, rtlClassName);
            pushToCSSRules(cssRulesByBucket, styleBucketName, ltrCSS, rtlCSS, atRules.media, computePropertyPriority(shorthand));
        } else if ((0, _isObject.isObject)(value)) {
            if ((0, _isNestedSelector.isNestedSelector)(property)) {
                resolveStyleRules(value, classNameHashSalt, selectors.concat((0, _normalizeNestedProperty.normalizeNestedProperty)(property)), atRules, cssClassesMap, cssRulesByBucket);
            } else if ((0, _isMediaQuerySelector.isMediaQuerySelector)(property)) {
                const combinedMediaQuery = (0, _generateCombinedMediaQuery.generateCombinedQuery)(atRules.media, property.slice(6).trim());
                resolveStyleRules(value, classNameHashSalt, selectors, {
                    ...atRules,
                    media: combinedMediaQuery
                }, cssClassesMap, cssRulesByBucket);
            } else if ((0, _isLayerSelector.isLayerSelector)(property)) {
                const combinedLayerQuery = (atRules.layer ? `${atRules.layer}.` : '') + property.slice(6).trim();
                resolveStyleRules(value, classNameHashSalt, selectors, {
                    ...atRules,
                    layer: combinedLayerQuery
                }, cssClassesMap, cssRulesByBucket);
            } else if ((0, _isSupportQuerySelector.isSupportQuerySelector)(property)) {
                const combinedSupportQuery = (0, _generateCombinedMediaQuery.generateCombinedQuery)(atRules.supports, property.slice(9).trim());
                resolveStyleRules(value, classNameHashSalt, selectors, {
                    ...atRules,
                    supports: combinedSupportQuery
                }, cssClassesMap, cssRulesByBucket);
            } else if ((0, _isContainerQuerySelector.isContainerQuerySelector)(property)) {
                // TODO implement nested container queries if needed
                // The only way to target multiple containers is to nest container queries
                // https://developer.mozilla.org/en-US/docs/Web/CSS/@container#nested_container_queries
                const containerQuery = property.slice(10).trim();
                resolveStyleRules(value, classNameHashSalt, selectors, {
                    ...atRules,
                    container: containerQuery
                }, cssClassesMap, cssRulesByBucket);
            } else if ((0, _isScopeSelector.isScopeSelector)(property)) {
                if (atRules.scope) {
                    if (process.env.NODE_ENV !== 'production') {
                        (0, _logError.logError)(`@griffel/react: nested "${property}" is not supported. ` + 'Only one @scope boundary can be applied to a rule; the inner @scope will be skipped.');
                    }
                    continue;
                }
                const scopeQuery = property.slice(6).trim();
                if (scopeQuery === '' || !scopeQuery.startsWith('to ')) {
                    if (process.env.NODE_ENV !== 'production') {
                        (0, _logError.logError)(`@griffel/react: "${property}" is not a supported @scope syntax. ` + 'Use "@scope to (SELECTOR)" to define a scope boundary. ' + 'The styles will be skipped.');
                    }
                    continue;
                }
                resolveStyleRules(value, classNameHashSalt, selectors, {
                    ...atRules,
                    scope: scopeQuery
                }, cssClassesMap, cssRulesByBucket);
            } else {
                (0, _warnAboutUnresolvedRule.warnAboutUnresolvedRule)(property, value);
            }
        }
    }
    return [
        cssClassesMap,
        cssRulesByBucket
    ];
} //# sourceMappingURL=resolveStyleRules.js.map
