// @apolo/shared-crm/src/mock/mock-xrm.js
/**
 * Mock del objeto Xrm de Power Platform para desarrollo local y testing.
 * Configura IDs de prueba que puedes sobrescribir por entorno.
 */
export const mockXrm = {
  Utility: {
    getGlobalContext: () => ({
      getClientUrl: () => 'http://localhost:5173',
      userSettings: {
        userId: '{00000000-0000-0000-0000-000000000001}',
      },
    }),
  },
  LoadFormContext: {
    getFormContext: () => ({
      data: {
        entity: {
          getId: () => '{00000000-0000-0000-0000-000000000002}',
        },
      },
      getAttribute: (name) => ({
        getValue: () => {
          const attrs = {
            accountid: '00000000-0000-0000-0000-000000000002',
          }
          return attrs[name] ?? null
        },
      }),
    }),
  },
  WebApi: {
    online: {
      retrieveRecord: async (entityName, id) => {
        console.warn(`[mockXrm] retrieveRecord(${entityName}, ${id}) - retornando null`)
        return null
      },
    },
  },
  Page: {
    getControl: (name) => ({
      setSrc: (url) => console.log(`[mockXrm] setSrc(${name}) → ${url}`),
    }),
    getAttribute: (name) => ({
      getValue: () => null,
      setValue: (val) => console.log(`[mockXrm] setAttribute(${name}) = ${val}`),
    }),
  },
}
