// @apolo/shared-utils/src/builders/cuenta-builders.js
import { formatoFecha } from '../formatting/date.js'
import { formatoNumeroCuentaTarjeta } from '../formatting/card-number.js'
import { isNullOrUndefinedOrEmpty } from '../validation/null-checks.js'

// NOTE: primeraLetraMayusculas viene de text.js, pero por dependencia circular la ponemos aquí también
import { primeraLetraMayusculas as _primeraLetraMayusculas } from '../formatting/text.js'

const estado_pago_null = {
  monto_facturado: null,
  total_abonos: null,
  pago_minimo: null,
  glosa: '',
}

const bloqueos_cuenta_null = {
  id_bloqueo_cuenta1: null,
  codigo_bloqueo_cuenta1: null,
  descripcion_bloqueo_cuenta1: null,
  id_bloqueo_cuenta2: null,
  codigo_bloqueo_cuenta2: null,
  descripcion_bloqueo_cuenta2: null,
}

/**
 * Construye el objeto de cuenta principal normalizado
 * @param {Object} data - Datos del API VistaActivos
 * @param {string} idCuenta
 * @param {boolean} visualizarDatosOfuscados
 * @param {Function} [desencriptarFn] - Función async para desencriptar (opcional)
 * @returns {Promise<Object>}
 */
export const construirCuentaPrincipal = async (data, idCuenta, visualizarDatosOfuscados, desencriptarFn = null) => {
  if (isNullOrUndefinedOrEmpty(data)) return null

  const numero_cuenta_tac_desencriptado =
    visualizarDatosOfuscados && !isNullOrUndefinedOrEmpty(data.numero_cuenta_tac) && desencriptarFn
      ? formatoNumeroCuentaTarjeta(await desencriptarFn(data.numero_cuenta_tac))
      : ''

  return {
    id_cuenta: isNullOrUndefinedOrEmpty(data.id_cuenta) ? idCuenta : data.id_cuenta,
    numero_cuenta: data.numero_cuenta,
    estado: data.estado_autorizador_cuenta.descripcion
      ? data.estado_autorizador_cuenta.descripcion.toLowerCase()
      : '',
    castigada: data.es_castigada,
    numero_cuenta_tac: data.numero_cuenta_tac,
    numero_cuenta_tac_desencriptado,
    bloqueos: {
      id_bloqueo_cuenta1: data.bloqueos_cuenta.id_tipo_bloqueo_cuenta_1,
      codigo_bloqueo_cuenta1: data.bloqueos_cuenta.codigo_bloqueo_cuenta_1,
      descripcion_bloqueo_cuenta1: data.bloqueos_cuenta.descripcion_bloqueo_cuenta_1,
      id_bloqueo_cuenta2: data.bloqueos_cuenta.id_tipo_bloqueo_cuenta_2,
      codigo_bloqueo_cuenta2: data.bloqueos_cuenta.codigo_bloqueo_cuenta_2,
      descripcion_bloqueo_cuenta2: data.bloqueos_cuenta.descripcion_bloqueo_cuenta_2,
    },
    afinidad: data.afinidad,
    local_apertura: data.local_apertura,
    fecha_apertura: !isNullOrUndefinedOrEmpty(data.fecha_apertura) ? formatoFecha(data.fecha_apertura) : null,
    estado_pago: isNullOrUndefinedOrEmpty(data.estado_pago)
      ? estado_pago_null
      : {
          monto_facturado: data.estado_pago.monto_facturado ?? null,
          total_abonos: data.estado_pago.total_abonos ?? null,
          pago_minimo: data.estado_pago.pago_minimo ?? null,
          glosa: !isNullOrUndefinedOrEmpty(data.estado_pago.glosa)
            ? _primeraLetraMayusculas(data.estado_pago.glosa)
            : '',
        },
    es_tarjeta: data.es_tarjeta,
    clave_web: data.posee_clave_web,
    clave_compra: data.posee_clave_compra,
    es_vigente: data.es_vigente ?? null,
    monto_facturado: data.estado_pago?.monto_facturado ?? null,
    es_pagado: data.es_pagado,
    fecha_vencimiento_facturacion: data.fecha_vencimiento_facturacion ?? null,
    fecha_proxima_facturacion: data.fecha_proxima_facturacion,
    dias_mora: data.dias_mora,
    cupo_total: data.cupo_nacional_cupo_virtual,
    cupo_utilizado: data.utilizado_calculado,
    cupo_disponible: data.dispo_total_en_clp,
    tiene_cupo_disponible: data.dispo_total_en_clp > 0,
    tiene_cupo_disponible_descripcion: data.dispo_total_en_clp > 0 ? 'Disponible' : 'Sin disponible',
    facturacion_ultimos_seis_meses: data.facturacion_ultimos_seis_meses,
  }
}

/**
 * Construye el objeto de cuenta adicional
 */
export const construirCuentaAdicional = (data) => {
  if (isNullOrUndefinedOrEmpty(data)) return null

  return {
    id_cuenta: data.cuenta.id_cuenta,
    numero_cuenta: data.cuenta.numero_cuenta,
    estado: data.cuenta.estado_autorizador_cuenta.descripcion
      ? data.cuenta.estado_autorizador_cuenta.descripcion.toLowerCase()
      : '',
    castigada: data.es_castigada,
    numero_cuenta_tac: data.numero_cuenta_tac,
    numero_cuenta_tac_desencriptado: '',
    bloqueos: {
      id_bloqueo_cuenta1: data.cuenta.bloqueos_cuenta.id_tipo_bloqueo_cuenta_1,
      codigo_bloqueo_cuenta1: data.cuenta.bloqueos_cuenta.codigo_bloqueo_cuenta_1,
      descripcion_bloqueo_cuenta1: data.cuenta.bloqueos_cuenta.descripcion_bloqueo_cuenta_1,
      id_bloqueo_cuenta2: data.cuenta.bloqueos_cuenta.id_tipo_bloqueo_cuenta_2,
      codigo_bloqueo_cuenta2: data.cuenta.bloqueos_cuenta.codigo_bloqueo_cuenta_2,
      descripcion_bloqueo_cuenta2: data.cuenta.bloqueos_cuenta.descripcion_bloqueo_cuenta_2,
    },
    afinidad: data.afinidad,
    local_apertura: 'Local Digital',
    fecha_apertura: data.fecha_apertura,
    estado_pago: isNullOrUndefinedOrEmpty(data.estado_pago)
      ? estado_pago_null
      : {
          monto_facturado: data.estado_pago.monto_facturado ?? null,
          total_abonos: data.estado_pago.total_abonos ?? null,
          pago_minimo: data.estado_pago.pago_minimo ?? null,
          glosa: !isNullOrUndefinedOrEmpty(data.estado_pago.glosa)
            ? _primeraLetraMayusculas(data.estado_pago.glosa)
            : '',
        },
    es_tarjeta: data.es_tarjeta,
    es_clave_compra: data.es_clave_compra,
    es_vigente: data.es_vigente ?? null,
    monto_facturado: data.estado_pago?.monto_facturado ?? null,
    es_pagado: data.es_pagado,
    fecha_vencimiento_facturacion: data.fecha_vencimiento_facturacion ?? null,
    fecha_proxima_facturacion: data.fecha_proxima_facturacion,
    dias_mora: data.dias_mora,
    cupo_total: data.cupo_nacional_cupo_virtual,
    cupo_utilizado: data.utilizado_calculado,
    cupo_disponible: data.dispo_total_en_clp,
    tiene_cupo_disponible: data.dispo_total_en_clp > 0,
    tiene_cupo_disponible_descripcion: data.dispo_total_en_clp > 0 ? 'Disponible' : 'Sin disponible',
    facturacion_ultimos_seis_meses: data.facturacion_ultimos_seis_meses,
    nombre_titular: data.cuenta.nombre_titular,
  }
}

/**
 * Construye cuenta no vigente
 */
export const construirCuentaNoVigente = (data, idCuenta) => {
  if (isNullOrUndefinedOrEmpty(data)) return null

  return {
    id_cuenta: idCuenta,
    numero_cuenta: data.numero_cuenta,
    estado: 'inactiva',
    numero_cuenta_tac: data.numero_cuenta_tac,
    numero_cuenta_tac_desencriptado: '',
    bloqueos: isNullOrUndefinedOrEmpty(data.bloqueo_cuentas)
      ? bloqueos_cuenta_null
      : {
          id_bloqueo_cuenta1: data.bloqueo_cuentas.id_tipo_bloqueo_cuenta_1 ?? null,
          codigo_bloqueo_cuenta1: data.bloqueo_cuentas.codigo_bloqueo_cuenta_1 ?? null,
          descripcion_bloqueo_cuenta1: data.bloqueo_cuentas.descripcion_bloqueo_cuenta_1 ?? null,
          id_bloqueo_cuenta2: data.bloqueo_cuentas.id_tipo_bloqueo_cuenta_2 ?? null,
          codigo_bloqueo_cuenta2: data.bloqueo_cuentas.codigo_bloqueo_cuenta_2 ?? null,
          descripcion_bloqueo_cuenta2: data.bloqueo_cuentas.descripcion_bloqueo_cuenta_2 ?? null,
        },
    afinidad: data.afinidad ?? null,
    local_apertura: data.local_apertura ?? null,
    fecha_apertura: !isNullOrUndefinedOrEmpty(data.fecha_apertura) ? formatoFecha(data.fecha_apertura) : null,
    estado_pago: isNullOrUndefinedOrEmpty(data.estado_pago)
      ? estado_pago_null
      : {
          monto_facturado: data.estado_pago.monto_facturado ?? null,
          total_abonos: data.estado_pago.total_abonos ?? null,
          pago_minimo: data.estado_pago.pago_minimo ?? null,
          glosa: data.estado_pago.glosa ?? null,
        },
    es_tarjeta: data.es_tarjeta ?? null,
    clave_web: data.posee_clave_web ?? null,
    clave_compra: data.posee_clave_compra ?? null,
    es_vigente: data.es_vigente ?? null,
    monto_facturado: data.estado_pago?.monto_facturado ?? null,
    es_pagado: data.es_pagado ?? null,
    fecha_vencimiento_facturacion: data.fecha_vencimiento_facturacion ?? null,
    fecha_proxima_facturacion: data.fecha_proxima_facturacion ?? null,
    dias_mora: data.dias_mora ?? null,
    cupo_total: data.cupo_nacional_cupo_virtual ?? null,
    cupo_utilizado: data.utilizado_calculado ?? null,
    cupo_disponible: data.dispo_total_en_clp ?? null,
    tiene_cupo_disponible: false,
    tiene_cupo_disponible_descripcion: 'Sin disponible',
    facturacion_ultimos_seis_meses: data.facturacion_ultimos_seis_meses ?? null,
  }
}
