// @apolo/shared-utils/src/builders/selector-tarjetas.js
import { ofuscarNumeroConAsteriscos } from '../formatting/card-number.js'

/**
 * Crea un item de selector para tarjetas pendientes de embozado
 */
export const new_selector_tarjetas_embozados = (embozado) => {
  return {
    codigo_logo: embozado.logo.codigo,
    numero_tarjeta: ofuscarNumeroConAsteriscos(embozado.numero_tarjeta),
    nombre_cliente: embozado.nombre_cliente,
    id_tarjeta: embozado.id_tarjeta,
    boton: {
      label: 'Embozar',
      icono: null,
      isNavigate: true,
      ruta: `/embozado/${embozado.id_tarjeta}`,
      metodo: null,
      disabled: false,
    },
    en_curso: false,
    background_color: '',
  }
}

/**
 * Crea un item de selector para tarjetas con embozado en curso
 */
export const new_selector_tarjetas_embozados_en_curso = (embozado) => {
  return {
    codigo_logo: embozado.CodigoLogo,
    numero_tarjeta: ofuscarNumeroConAsteriscos(embozado.NroTarjeta),
    nombre_cliente: embozado.NombrePersona,
    id_tarjeta: embozado.IdTarjeta,
    boton: {
      label: 'Embozado en curso',
      icono: null,
      isNavigate: true,
      ruta: `/embozado/${embozado.IdTarjeta}`,
      metodo: null,
      disabled: false,
    },
    en_curso: true,
    background_color: 'en-curso',
  }
}
