// @apolo/shared-utils/src/builders/tarjeta-builders.js
import { formatoFecha } from '../formatting/date.js'
import { isNullOrUndefinedOrEmpty } from '../validation/null-checks.js'
import { formatoNumeroCuentaTarjeta } from '../formatting/card-number.js'

const bloqueos_tarjeta_null = {
  id: null,
  codigo: null,
  descripcion: null,
  permite_compra: null,
  fecha_hora_bloqueo: null,
}

/**
 * Construye la tarjeta principal desde datos de VistaActivos
 */
export const construirTarjetaPrincipal = async (data, visualizarDatosOfuscados, desencriptarFn = null) => {
  if (isNullOrUndefinedOrEmpty(data)) return null

  const filtrarTarjetaPrincipal = data.tarjetas?.find((t) => t.es_principal === true)
  if (isNullOrUndefinedOrEmpty(filtrarTarjetaPrincipal)) return null

  const t = filtrarTarjetaPrincipal
  const numero_tarjeta_tac_desencriptado =
    visualizarDatosOfuscados && !isNullOrUndefinedOrEmpty(t.numero_tarjeta_tac) && desencriptarFn
      ? formatoNumeroCuentaTarjeta(await desencriptarFn(t.numero_tarjeta_tac))
      : ''

  return {
    id_tarjeta: t.id_tarjeta,
    numero_tarjeta: t.numero_tarjeta,
    numero_tarjeta_tac: t.numero_tarjeta_tac,
    numero_tarjeta_tac_desencriptado,
    logo: { id: t.logo.id, codigo: t.logo.codigo, descripcion: t.logo.descripcion },
    tarjeta_digital: t.es_digital,
    firma: !data.firma,
    estado: t.estado_autorizador_tarjeta.descripcion
      ? t.estado_autorizador_tarjeta.descripcion.toLowerCase()
      : '',
    expiracion: formatoFecha(t.fecha_vencimiento_tarjeta),
    bloqueo: {
      id: t.bloqueo_tarjeta?.id ?? null,
      codigo: t.bloqueo_tarjeta?.codigo ?? null,
      descripcion: t.bloqueo_tarjeta?.descripcion ?? null,
      permite_compra: t.bloqueo_tarjeta?.permite_compra ?? null,
      fecha_hora_bloqueo: t.bloqueo_tarjeta?.fecha_hora_bloqueo ?? null,
    },
    fecha_emision: formatoFecha(t.fecha_emision),
    fecha_embozado: formatoFecha(t.fecha_embozado),
    estado_tokenizacion: t.estado_tokenizacion,
    clave_web: t.es_digital,
    clave_compra: t.es_clave_compra,
    es_vigente: t.es_vigente ?? null,
    icono_on: data.icono_on ?? null,
    icono_off: data.icono_off ?? null,
  }
}

/**
 * Construye tarjeta adicional
 */
export const construirTarjetaAdicional = (data) => {
  if (isNullOrUndefinedOrEmpty(data)) return null

  return {
    id_tarjeta: data.id_tarjeta,
    numero_tarjeta: data.numero_tarjeta,
    numero_tarjeta_tac: data.numero_tarjeta_tac,
    numero_tarjeta_tac_desencriptado: '',
    logo: { id: data.logo.id, codigo: data.logo.codigo, descripcion: data.logo.descripcion },
    tarjeta_digital: data.es_digital,
    firma: data.firma,
    estado: data.estado_autorizador_tarjeta.descripcion
      ? data.estado_autorizador_tarjeta.descripcion.toLowerCase()
      : '',
    expiracion: formatoFecha(data.fecha_vencimiento_tarjeta),
    bloqueo: {
      id: data.bloqueos_tarjeta?.id ?? null,
      codigo: data.bloqueos_tarjeta?.codigo ?? null,
      descripcion: data.bloqueos_tarjeta?.descripcion ?? null,
      permite_compra: data.bloqueos_tarjeta?.permite_compra ?? null,
      fecha_hora_bloqueo: data.bloqueos_tarjeta?.fecha_hora_bloqueo ?? null,
    },
    fecha_emision: formatoFecha(data.fecha_emision),
    fecha_embozado: formatoFecha(data.fecha_embozado),
    estado_tokenizacion: data.estado_tokenizacion,
    clave_web: data.posee_clave_web,
    clave_compra: data.posee_clave_compra,
    es_vigente: data.es_vigente,
    icono_on: data.icono_on ?? null,
    icono_off: data.icono_off ?? null,
  }
}

/**
 * Construye tarjeta no vigente
 */
export const construirTarjetaNoVigente = (data) => {
  if (isNullOrUndefinedOrEmpty(data)) return null

  return {
    id_tarjeta: data.id_tarjeta ?? null,
    numero_tarjeta: data.numero_tarjeta ?? null,
    numero_tarjeta_tac: data.numero_tarjeta_tac ?? null,
    numero_tarjeta_tac_desencriptado: '',
    logo: {
      id: data.logo?.id ?? null,
      codigo: data.logo?.codigo ?? null,
      descripcion: data.logo?.descripcion ?? null,
    },
    tarjeta_digital: data.es_digital ?? null,
    firma: false,
    estado: 'inactiva',
    expiracion: !isNullOrUndefinedOrEmpty(data.fecha_vencimiento_tarjeta)
      ? formatoFecha(data.fecha_vencimiento_tarjeta)
      : null,
    bloqueo: isNullOrUndefinedOrEmpty(data.bloqueo_tarjeta)
      ? bloqueos_tarjeta_null
      : {
          id: data.bloqueo_tarjeta?.id ?? null,
          codigo: data.bloqueo_tarjeta?.codigo ?? null,
          descripcion: data.bloqueo_tarjeta?.descripcion ?? null,
          permite_compra: data.bloqueo_tarjeta?.permite_compra ?? null,
          fecha_hora_bloqueo: data.bloqueo_tarjeta?.fecha_hora_bloqueo ?? null,
        },
    fecha_emision: !isNullOrUndefinedOrEmpty(data.fecha_emision) ? formatoFecha(data.fecha_emision) : null,
    fecha_embozado: !isNullOrUndefinedOrEmpty(data.fecha_embozado) ? formatoFecha(data.fecha_embozado) : null,
    estado_tokenizacion: data.estado_tokenizacion ?? null,
    clave_web: data.es_digital ?? null,
    clave_compra: data.es_clave_compra ?? null,
    es_vigente: data.es_vigente ?? null,
    icono_on: data.icono_on ?? null,
    icono_off: data.icono_off ?? null,
  }
}
