// @fluentui/react-positioning/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    POSITIONING_SLIDE_DIRECTION_VAR_X: function() {
        return _constants.POSITIONING_SLIDE_DIRECTION_VAR_X;
    },
    POSITIONING_SLIDE_DIRECTION_VAR_Y: function() {
        return _constants.POSITIONING_SLIDE_DIRECTION_VAR_Y;
    },
    PositioningConfigurationProvider: function() {
        return _PositioningConfigurationContext.PositioningConfigurationProvider;
    },
    createArrowHeightStyles: function() {
        return _createArrowStyles.createArrowHeightStyles;
    },
    createArrowStyles: function() {
        return _createArrowStyles.createArrowStyles;
    },
    createSlideStyles: function() {
        return _createSlideStyles.createSlideStyles;
    },
    createVirtualElementFromClick: function() {
        return _createVirtualElementFromClick.createVirtualElementFromClick;
    },
    mergeArrowOffset: function() {
        return _index.mergeArrowOffset;
    },
    resolvePositioningShorthand: function() {
        return _index.resolvePositioningShorthand;
    },
    usePositioning: function() {
        return _usePositioning.usePositioning;
    },
    usePositioningMouseTarget: function() {
        return _usePositioningMouseTarget.usePositioningMouseTarget;
    },
    usePositioningSlideDirection: function() {
        return _usePositioningSlideDirection.usePositioningSlideDirection;
    },
    useSafeZoneArea: function() {
        return _useSafeZoneArea.useSafeZoneArea;
    }
});
const _createVirtualElementFromClick = require("./createVirtualElementFromClick");
const _usePositioningSlideDirection = require("./usePositioningSlideDirection");
const _constants = require("./constants");
const _createArrowStyles = require("./createArrowStyles");
const _createSlideStyles = require("./createSlideStyles");
const _PositioningConfigurationContext = require("./PositioningConfigurationContext");
const _usePositioning = require("./usePositioning");
const _usePositioningMouseTarget = require("./usePositioningMouseTarget");
const _useSafeZoneArea = require("./hooks/useSafeZoneArea/useSafeZoneArea");
const _index = require("./utils/index");
