// @griffel/core/lib/common/snapshotSerializers.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get griffelRendererSerializer () {
        return griffelRendererSerializer;
    },
    get griffelResetRulesSerializer () {
        return griffelResetRulesSerializer;
    },
    get griffelRulesSerializer () {
        return griffelRulesSerializer;
    }
});
const _prettier = /*#__PURE__*/ _interop_require_wildcard(require("prettier"));
const _constants = require("../constants.cjs");
const _normalizeCSSBucketEntry = require("../runtime/utils/normalizeCSSBucketEntry.cjs");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
// eslint-disable-next-line eqeqeq
const isObject = (value)=>value != null && !Array.isArray(value) && typeof value === 'object';
const griffelRendererSerializer = {
    test (value) {
        return isObject(value) && isObject(value.stylesheets);
    },
    print (value) {
        /**
         * test function makes sure that value is the guarded type
         */ const _value = value;
        const stylesheetKeys = Object.keys(_value.stylesheets);
        const rules = stylesheetKeys.reduce((acc, styleEl)=>{
            var _a;
            const stylesheet = _value.stylesheets[styleEl];
            if (stylesheet) {
                const cssRules = (_a = stylesheet.cssRules()) !== null && _a !== void 0 ? _a : [];
                const attributes = Object.entries(stylesheet.elementAttributes).filter(([key])=>key !== _constants.DATA_BUCKET_ATTR);
                if (cssRules.length === 0) {
                    return acc;
                }
                return [
                    ...acc,
                    `/** bucket "${styleEl.slice(0, 1)}"${attributes.length > 0 ? ' ' + JSON.stringify(Object.fromEntries(attributes)) : ''} **/`,
                    ...cssRules
                ];
            }
            return acc;
        }, []);
        return _prettier.format(rules.join('\n'), {
            parser: 'css'
        }).trim();
    }
};
const isStyleRulesTuple = (value)=>Array.isArray(value) && value.length === 2;
const griffelRulesSerializer = {
    test (value) {
        return isStyleRulesTuple(value);
    },
    print (value) {
        /**
         * test function makes sure that value is the guarded type
         */ const _value = value;
        const cssRulesByBucket = _value[1];
        const keys = Object.keys(cssRulesByBucket);
        return keys.reduce((acc, styleBucketName)=>{
            const rules = cssRulesByBucket[styleBucketName].map((entry)=>(0, _normalizeCSSBucketEntry.normalizeCSSBucketEntry)(entry)[0]);
            return _prettier.format(acc + rules.join(''), {
                parser: 'css'
            }).trim();
        }, '');
    }
};
const griffelResetRulesSerializer = {
    test (value) {
        return Array.isArray(value);
    },
    print (value) {
        const format = (cssRules)=>_prettier.format(cssRules.join(''), {
                parser: 'css'
            }).trim();
        /**
         * test function makes sure that value is the guarded type
         */ const _value = value;
        const cssRulesByBucket = Array.isArray(_value[2]) ? {
            r: _value[2]
        } : _value[2];
        return Object.entries(cssRulesByBucket).filter(([_key, value])=>value.length > 0).flatMap(([key, value])=>[
                `/** bucket "${key}" */`,
                format(value)
            ]).join('\n');
    }
}; //# sourceMappingURL=snapshotSerializers.js.map
