// @fluentui/react-card/lib-commonjs/components/CardFooter/useCardFooter.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useCardFooterBase_unstable: function() {
        return useCardFooterBase_unstable;
    },
    useCardFooter_unstable: function() {
        return useCardFooter_unstable;
    }
});
const _reactutilities = require("@fluentui/react-utilities");
const useCardFooter_unstable = (props, ref)=>{
    return useCardFooterBase_unstable(props, ref);
};
const useCardFooterBase_unstable = (props, ref)=>{
    const { action } = props;
    return {
        components: {
            root: 'div',
            action: 'div'
        },
        root: _reactutilities.slot.always((0, _reactutilities.getIntrinsicElementProps)('div', {
            // FIXME:
            // `ref` is wrongly assigned to be `HTMLElement` instead of `HTMLDivElement`
            // but since it would be a breaking change to fix it, we are casting ref to it's proper type
            ref: ref,
            ...props
        }), {
            elementType: 'div'
        }),
        action: _reactutilities.slot.optional(action, {
            elementType: 'div'
        })
    };
};
